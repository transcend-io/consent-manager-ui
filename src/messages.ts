/* eslint-disable max-lines */
import { defineMessages } from '@transcend-io/internationalization';

export const messages = defineMessages('ui.src.messages', {
  consentTitle: {
    defaultMessage: 'What can we use data for?',
    description:
      'The title displayed when asking the user for specific opt out/in preferences.',
  },
  consentTitleAcceptAll: {
    defaultMessage: 'We use cookies',
    description:
      'The title displayed when asking the user to accept all data processing.',
  },
  completeOptionsInvertedTitle: {
    defaultMessage: 'Your Privacy Choices',
    description:
      'The title displayed in the CompleteOptionsInverted view state.',
  },
  acceptAllDescription: {
    defaultMessage:
      // eslint-disable-next-line max-len
      'By clicking “Accept all”, you agree to the storing of cookies on your device for functional, analytics, and advertising purposes.',
    description:
      'The description displayed when asking the user to accept all data processing.',
  },
  consentTitlePrivacyPolicyNotice: {
    defaultMessage: 'Our Privacy Policy Has Changed',
    description: 'The title displayed for the PrivacyPolicyNotice banner.',
  },
  privacyPolicyNoticeDescription: {
    defaultMessage:
      "We've updated your privacy policy. Click the link below to learn more about how we use your personal information.",
    description:
      'The description displayed for the PrivacyPolicyNotice banner.',
  },
  consentTitlePrivacyPolicyNoticeWithCloseButton: {
    defaultMessage: 'Our Privacy Policy Has Changed',
    description:
      'The title displayed for the PrivacyPolicyNoticeWithCloseButton banner.',
  },
  privacyPolicyNoticeWithCloseButtonDescription: {
    defaultMessage:
      "We've updated your privacy policy. Click the link below to learn more about how we use your personal information.",
    description:
      'The description displayed for the PrivacyPolicyNoticeWithCloseButton banner.',
  },
  consentTitleDoNotSellExplainer: {
    defaultMessage: 'Do Not Sell or Share My Personal Information',
    description: 'The title displayed for the DoNotSellExplainer banner.',
  },
  consentTitleAcceptOrRejectAnalytics: {
    defaultMessage: 'This website uses analytics',
    description: 'The title displayed for the AcceptOrRejectAnalytics banner.',
  },
  acceptAnalytics: {
    defaultMessage: 'Okay',
    description:
      'The affirmative response for opting into analytics in the "AcceptOrRejectAnalytics" view state.',
  },
  rejectAnalytics: {
    defaultMessage: `Do not track`,
    description:
      'The negative response for opting out of analytics in the "AcceptOrRejectAnalytics" view state.',
  },
  acceptOrRejectAnalyticsDescription: {
    defaultMessage:
      'By clicking "Okay", you agree to the usage of your sensitive information for analytics purposes.',
    description:
      'The description displayed for the AcceptOrRejectAnalytics banner.',
  },
  consentTitleAcceptOrRejectAdvertising: {
    defaultMessage: 'This website uses advertising',
    description:
      'The title displayed for the AcceptOrRejectAdvertising banner.',
  },
  acceptAdvertising: {
    defaultMessage: 'Okay',
    description:
      'The affirmative response for opting into advertising in the "AcceptOrRejectAdvertising" view state.',
  },
  rejectAdvertising: {
    defaultMessage: `Opt out of advertising`,
    description:
      'The negative response for opting out of advertising in the "AcceptOrRejectAdvertising" view state.',
  },
  acceptOrRejectAdvertisingDescription: {
    defaultMessage:
      'By clicking "Okay", you agree to the usage of your personal information for advertising purposes.',
    description:
      'The description displayed for the AcceptOrRejectAdvertising banner.',
  },
  privacyPolicyNoticeButton: {
    defaultMessage: 'Okay',
    description: 'Button to dismiss privacy policy notice.',
  },
  doNotSellOptedOut: {
    defaultMessage: `You have been successfully opted out.`,
    description:
      'Explainer text in DoNotSellExplainer banner when opted out of sale of data.',
  },
  close: {
    defaultMessage: 'Close consent manager',
    description: 'An accessible label for the button to close the modal',
  },
  switchLanguage: {
    defaultMessage: 'Switch language',
    description: 'An accessible label for the button to switch language',
  },
  doNotSellOptedIn: {
    defaultMessage: `Switch the toggle to opt out.`,
    description:
      'Explainer text in DoNotSellExplainer banner when opted in to sale of data.',
  },
  doNotSellDescription: {
    /* eslint-disable max-len */
    defaultMessage: `<p>We do not disclose your personal information to third parties in exchange for monetary consideration. In some instances, however, we may disclose your personal information to third parties in exchange for other valuable consideration, such as to enhance our product offerings, offer display advertisements, and in other ways which you can read about in our Privacy Statement.</p>
    <p>By opting out below, we will not disclose your personal information to third parties in exchange for valuable consideration.</p>
    <p>Please note that by opting out of these types of disclosures, you may limit our ability to customize your experience with content that may be of interest to you.</p>
    <p>If you have an account with us, sign in to persist your opt-out.</p>
    <p>If you do not have an account, you may opt-out of the disclosure of your personal information to third parties for valuable consideration (e.g., information stored in cookies) from this web browser below.</p>`,
    /* eslint-enable max-len */
    description: 'The description text for the the DoNotSellExplainer banner.',
  },
  noticeTitle: {
    defaultMessage: 'This website processes personal data',
    description: 'The Do Not Sell/Share notice title.',
  },
  saving: {
    defaultMessage: 'Saving...',
    description: 'The callback when preferences are being persisted.',
  },
  preferencesSaved: {
    defaultMessage: 'Preferences Saved!',
    description:
      'Confirmation that preferences are saved to an opted out state.',
  },
  preferencesSavedOptedIn: {
    defaultMessage: 'Preferences Saved!',
    description:
      'Confirmation that preferences are saved to an opted in state.',
  },
  collapsedLabel: {
    defaultMessage: 'Privacy Settings',
    description:
      'The title displayed in the collapsed version of the consent manager.',
  },
  acceptAllButtonPrimary: {
    defaultMessage: 'Accept all',
    description: 'Button text for accepting all trackers.',
  },
  rejectAllButtonPrimary: {
    defaultMessage: 'Reject all',
    description: 'Button text for rejecting all trackers.',
  },
  moreChoicesButtonPrimary: {
    defaultMessage: 'More choices',
    description:
      'Button text for redirecting the user to more granular consent choices.',
  },
  consentTitleAcceptAllRejectAllToggle: {
    defaultMessage: 'Your Privacy Choices',
    description:
      'The title displayed in the AcceptAllRejectAllToggle view state.',
  },
  acceptAllRejectAllToggleDescription: {
    defaultMessage:
      /* eslint-disable max-len */
      'By opting in below, you agree to the storing of cookies on your device for functional, analytics, and advertising purposes.',
    /* eslint-enable max-len */
    description:
      'The description text for the AcceptAllRejectAllToggle view state.',
  },
  consentTitleCompleteOptionsToggle: {
    defaultMessage: 'Privacy Settings',
    description: 'The title text for the CompleteOptionsToggles view state.',
  },
  consentTitleCompleteOptionsToggleDescription: {
    defaultMessage:
      'Below is an overview of the different technologies we use when processing your personal information.',
    description:
      'The description text for the CompleteOptionsToggles view state.',
  },
  modalAriaLabel: {
    defaultMessage: 'User data usage dialog',
    description:
      'The aria label used by screen readers for the modal. This is read when the modal is opened and focused.',
  },
  modalAriaDescription: {
    defaultMessage:
      'Please review the below information or selections regarding your user data usage.',
    description:
      'The aria description used by screen readers for the modal. This is read when the modal is opened and focused.',
  },
  acceptAllButtonAriaDescription: {
    defaultMessage: 'Accept all cookies and close dialog.',
    description:
      'The aria description read when the accept all button is focused.',
  },
  noticeButtonAriaDescription: {
    defaultMessage: 'Acknowledge and close dialog.',
    description: 'The aria description read when the ok button is focused.',
  },
  buttonGroupAriaDescription: {
    defaultMessage: 'Select consent and close dialog',
    description: 'Description for consent submission button groups.',
  },
});

export const quickOptionsMessages = defineMessages('ui.src.quickOptions', {
  essentialsButtonPrimary: {
    defaultMessage: 'Only Essentials',
    description:
      'Primary button text for selecting only essential data flows/cookies.',
  },
  essentialsButtonSecondary: {
    defaultMessage: 'required to operate',
    description:
      'Secondary button text for selecting only essential data flows/cookies.',
  },
  functionalButtonPrimary: {
    defaultMessage: 'Functional',
    description:
      'Primary button text for selecting functional and essential data flows/cookies.',
  },
  functionalButtonSecondary: {
    defaultMessage: 'extra site features',
    description:
      'Secondary button text for selecting functional and essential data flows/cookies.',
  },
  functionalAnalyticsButtonPrimary: {
    defaultMessage: 'Analytics',
    description:
      'Primary button text for selecting functional, analytics, and essential data flows/cookies.',
  },
  functionalAnalyticsButtonSecondary: {
    defaultMessage: 'extra site features',
    description:
      'Secondary button text for selecting functional, analytics, and essential data flows/cookies.',
  },
  analyticsButtonPrimary: {
    defaultMessage: 'Analytics',
    description:
      'Primary button text for selecting functional, essential, and analytics data flows/cookies.',
  },
  analyticsButtonSecondary: {
    defaultMessage: 'helps us improve',
    description:
      'Secondary button text for selecting functional, essential, and analytics data flows/cookies.',
  },
  advertisingButtonPrimary: {
    defaultMessage: 'Ads and Analytics',
    description:
      'Primary button text for selecting functional, essential, analytics and advertising data flows/cookies.',
  },
  advertisingButtonSecondary: {
    defaultMessage: 'helps us personalize ads',
    description:
      'Secondary button text for selecting functional, essential, analytics and advertising data flows/cookies.',
  },
});

export const noticeAndDoNotSellMessages = defineMessages(
  'ui.src.noticeAndDoNotSell',
  {
    confirmButtonPrimary: {
      defaultMessage: 'Okay',
      description: 'OK button text for the Do Not Sell/Share experience.',
    },
    doNotSellPrimary: {
      defaultMessage: 'Do Not Sell My Personal Information',
      description:
        'Do Not Sell notice in the footer of the Do Not Sell/Share experience.',
    },
    doNotSellLabel: {
      defaultMessage: 'Opt out of the sale of your personal information',
      description:
        'Hover/alt text for Do Not Sell notice in footer of the Do Not Sell/Share experience.',
    },
    doNotSellHonored: {
      defaultMessage: 'Your opt-out request has been honored.',
      description:
        'Title message for acknowledging do not sell preference has ben honored through click.',
    },
    doNotSellHonoredDescription: {
      defaultMessage:
        // eslint-disable-next-line max-len
        "We've received your request for this device or browser, and we will no longer sell or share your data with third parties. Learn more about our data practices in our privacy policy.",
      description:
        'Description message for acknowledging do not sell preference has ben honored through click.',
    },
    doNotSellHonoredGpc: {
      defaultMessage: 'Opt-out preference signal honored.',
      description:
        'Title message for acknowledging do not sell preference has ben honored through GPC.',
    },
  },
);

export const optOutDisclosureMessages = defineMessages(
  'ui.src.optOutDisclosure',
  {
    optOutHonored: {
      defaultMessage: 'Your opt-out request has been honored.',
      description:
        'Title message for acknowledging opt out preference has ben honored through 1-click option.',
    },
    optOutHonoredDescription: {
      defaultMessage:
        // eslint-disable-next-line max-len
        "We've received your request for this device or browser, and you have been opted out. Learn more about our data practices in our privacy policy.",
      description:
        'Description message for acknowledging opt out preference has ben honored through 1-click option.',
    },
    optOutHonoredGpc: {
      defaultMessage: 'Opt-out preference signal honored.',
      description:
        'Title message for acknowledging do not sell preference has ben honored through GPC.',
    },
  },
);

export const bottomMenuMessages = defineMessages('ui.src.bottomMenu', {
  backButtonText: {
    defaultMessage: 'Go back',
    description:
      'Hover text the back button in the footer of the consent banner language options.',
  },
  backButtonTooltip: {
    defaultMessage: 'Go back',
    description:
      'Main text for the back button in the footer of the consent banner language options.',
  },
  moreChoicesButtonPrimary: {
    defaultMessage: 'More choices',
    description: 'Text for selecting specific more opt out choices.',
  },
  moreChoicesButtonLabel: {
    defaultMessage: 'Click to show more choices',
    description: 'Hover/alt text for selecting specific privacy choices.',
  },
  simplerChoicesButtonPrimary: {
    defaultMessage: 'Simpler choices',
    description: 'Text for selecting simpler opt out choices.',
  },
  simplerChoicesButtonLabel: {
    defaultMessage: 'Click to show simpler choices',
    description: 'Hover/alt text for selecting simpler opt out choices.',
  },
  showPolicyButtonLanguageOptions: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the LanguageOptions UI.',
  },
  showPolicyButtonDoNotSellDisclosure: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the DoNotSellDisclosure UI.',
  },
  showPolicyButtonOptOutDisclosure: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in OptOutDisclosure UI.',
  },
  showPolicyButtonQuickOptions: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the QuickOptions UI.',
  },
  showPolicyButtonQuickOptions3: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the QuickOptions3 UI.',
  },
  showPolicyButtonAcceptAll: {
    defaultMessage: 'See our privacy policy',
    description: 'Text for linking out to privacy policy in the AcceptAll UI.',
  },
  showPolicyButtonAcceptAllOrMoreChoices: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the AcceptAllOrMoreChoices UI.',
  },
  showPolicyButtonCompleteOptionsToggles: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the CompleteOptionsToggles UI.',
  },
  showPolicyButtonAcceptOrRejectAll: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the AcceptOrRejectAll UI.',
  },
  showPolicyButtonAcceptOrRejectAllOrMoreChoices: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the AcceptOrRejectAllOrMoreChoices UI.',
  },
  showPolicyButtonAcceptOrRejectAnalytics: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the AcceptOrRejectAnalytics UI.',
  },
  showPolicyButtonAcceptOrRejectAdvertising: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the AcceptOrRejectAdvertising UI.',
  },
  showPolicyButtonAcceptAllRejectAllToggle: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the AcceptAllRejectAllToggle UI.',
  },
  showPolicyButtonNoticeAndDoNotSell: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the NoticeAndDoNotSell UI.',
  },
  showPolicyButtonDoNotSellExplainer: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the DoNotSellExplainer UI.',
  },
  showPolicyButtonPrivacyPolicyNotice: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the PrivacyPolicyNotice UI.',
  },
  showPolicyButtonPrivacyPolicyNoticeWithCloseButton: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the PrivacyPolicyNoticeWithCloseButton UI.',
  },
  showPolicyButtonCompleteOptions: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the CompleteOptions UI.',
  },
  showPolicyButtonCompleteOptionsInverted: {
    defaultMessage: 'See our privacy policy',
    description:
      'Text for linking out to privacy policy in the CompleteOptionsInverted UI.',
  },
  // END --- Privacy Policy Link Text, per ViewState
  showPolicyButtonLabel: {
    defaultMessage: 'Visit our privacy policy',
    description: 'Hover/alt for linking out to privacy policy.',
  },
  showSecondaryPolicyButton: {
    defaultMessage: 'Read more',
    description: 'Text for linking out to secondary policy.',
  },
  showSecondaryPolicyButtonLabel: {
    defaultMessage: 'Read more about your privacy choices',
    description: 'Hover/alt for linking out to secondary policy.',
  },
});

export const completeOptionsMessages = defineMessages(
  'ui.src.completeOptions',
  {
    essentialLabel: {
      defaultMessage: 'Essential purposes',
      description: 'Text for essential purposes in CompleteOptions view state.',
    },
    functionalLabel: {
      defaultMessage: 'Functionality',
      description:
        'Text for functional purposes in CompleteOptions view state.',
    },
    analyticsLabel: {
      defaultMessage: 'Analytics',
      description: 'Text for analytics purposes in CompleteOptions view state.',
    },
    advertisingLabel: {
      defaultMessage: 'Advertising',
      description:
        'Text for advertising purposes in CompleteOptions view state.',
    },
    saleOfInfoLabel: {
      defaultMessage: 'Sale of personal information',
      description:
        'Text for sale of information purposes in CompleteOptions view state.',
    },
    saveButtonPrimary: {
      defaultMessage: 'Confirm',
      description: 'Confirm button text in CompleteOptions view state.',
    },
    essentialAriaLabel: {
      defaultMessage: 'Essential data collection cannot be turned off.',
      description:
        'Hover/alt text for disabled essential purposes in CompleteOptions view state.',
    },
    toggleDisable: {
      defaultMessage: 'Click to disable collection of data for this purpose',
      description:
        'Hover/alt text for disabled opt in/opt out checkboxes when user is opted in in CompleteOptions view state.',
    },
    toggleEnable: {
      defaultMessage: 'Click to enable collection of data for this purpose',
      description:
        'Hover/alt text for disabled opt in/opt out checkboxes when user is opted out in CompleteOptions view state.',
    },
    globalPrivacySignal: {
      defaultMessage:
        "Set automatically by your browser's Global Privacy Control signal.",
      description:
        'Text to display when Global Privacy Control (GPC) is respected.',
    },
    description: {
      defaultMessage: '-',
      description:
        'Freeform description text for the CompleteOptions consent banner.',
    },
  },
);

export const completeOptionsInvertedMessages = defineMessages(
  'ui.src.completeOptionsInverted',
  {
    description: {
      defaultMessage: '-',
      description:
        'Freeform description text for the CompleteOptionsInverted consent banner.',
    },
    functionalLabel: {
      defaultMessage: 'Opt out of Functional',
      description:
        'Text for functional purposes in CompleteOptionsInverted view state.',
    },
    analyticsLabel: {
      defaultMessage: 'Opt out of Analytics',
      description:
        'Text for analytics purposes in CompleteOptionsInverted view state.',
    },
    advertisingLabel: {
      defaultMessage: 'Opt out of Advertising',
      description:
        'Text for advertising purposes in CompleteOptionsInverted view state.',
    },
    saleOfInfoLabel: {
      defaultMessage: 'Do Not Sell/Share My Personal Information.',
      description:
        'Text for sale of information purposes in CompleteOptionsInverted view state.',
    },
  },
);
/* eslint-enable max-lines */
