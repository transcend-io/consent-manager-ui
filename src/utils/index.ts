export * from './comma-and-or-space-separated-list';
export * from './compile';
export * from './create-html-element';
export * from './create-html-element';
export * from './get-user-locales';
export * from './getTranscendPolicies';
export * from './idle-callback';
export * from './safe-json-parse';
export * from './sort-locales-by-preference';
export * from './substitute-html';
export * from './throw-outside';
