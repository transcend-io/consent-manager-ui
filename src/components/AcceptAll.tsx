import { h, JSX } from 'preact';
import { useIntl } from 'react-intl';
import { useAirgap } from '../hooks';
import { messages } from '../messages';
import type { HandleSetViewState } from '../types';
import { Button } from './Button';

/**
 * Component showing "accept all" interface
 */
export function AcceptAll({
  handleSetViewState,
}: {
  /** Function to change viewState */
  handleSetViewState: HandleSetViewState;
}): JSX.Element {
  const { airgap } = useAirgap();
  const { formatMessage } = useIntl();

  // Opt in to all purposes
  const handleAcceptAll:
    | JSX.MouseEventHandler<HTMLButtonElement>
    | undefined = (
    event: JSX.TargetedEvent<HTMLButtonElement, MouseEvent>,
  ): void => {
    event.preventDefault();
    airgap.optIn(event);
    handleSetViewState('close');
  };

  return (
    <div className="column-content">
      <div>
        <p className="text-title text-title-left">
          {formatMessage(messages.consentTitleAcceptAll)}
        </p>
        <p className="paragraph">
          {formatMessage(messages.acceptAllDescription)}
        </p>
      </div>
      <Button
        primaryText={formatMessage(messages.acceptAllButtonPrimary)}
        handleClick={handleAcceptAll}
      />
    </div>
  );
}
