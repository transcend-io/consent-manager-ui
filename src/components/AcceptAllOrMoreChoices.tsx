import { h, JSX } from 'preact';
import { useIntl } from 'react-intl';
import { useAirgap } from '../hooks';
import { messages } from '../messages';
import type { HandleSetViewState } from '../types';
import { Button } from './Button';

/**
 * Component showing "accept all or more choices" interface
 */
export function AcceptAllOrMoreChoices({
  handleSetViewState,
}: {
  /** Function to change viewState */
  handleSetViewState: HandleSetViewState;
}): JSX.Element {
  const { airgap } = useAirgap();
  const { formatMessage } = useIntl();

  // Opt in to all purposes
  const handleAcceptAll:
    | JSX.MouseEventHandler<HTMLButtonElement>
    | undefined = (
    event: JSX.TargetedEvent<HTMLButtonElement, MouseEvent>,
  ): void => {
    event.preventDefault();
    airgap.optIn(event);
    handleSetViewState('close');
  };

  // redirect to more choices
  const handleMoreChoices:
    | JSX.MouseEventHandler<HTMLButtonElement>
    | undefined = (
    event: JSX.TargetedEvent<HTMLButtonElement, MouseEvent>,
  ): void => {
    event.preventDefault();
    handleSetViewState('CompleteOptions');
  };

  return (
    <div className="column-content" role="none">
      <div>
        <div>
          <p
            id="consent-dialog-title"
            role="heading"
            className="text-title text-title-left"
          >
            {formatMessage(messages.consentTitleAcceptAll)}
          </p>
        </div>
        <div>
          <p className="paragraph">
            <div
              role="paragraph"
              // eslint-disable-next-line react/no-danger
              dangerouslySetInnerHTML={{
                __html: formatMessage(messages.acceptAllDescription),
              }}
            />
          </p>
        </div>
      </div>
      <div
        className="accept-or-reject-all-button-row"
        role="group"
        aria-label={formatMessage(messages.buttonGroupAriaDescription)}
      >
        <Button
          primaryText={formatMessage(messages.moreChoicesButtonPrimary)}
          handleClick={handleMoreChoices}
          initialFocus
        />
        <Button
          primaryText={formatMessage(messages.acceptAllButtonPrimary)}
          handleClick={handleAcceptAll}
        />
      </div>
    </div>
  );
}
