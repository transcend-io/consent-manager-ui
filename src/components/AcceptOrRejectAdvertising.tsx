import { h, JSX } from 'preact';
import { useIntl } from 'react-intl';
import { CONSENT_OPTIONS } from '../constants';
import { useAirgap } from '../hooks';
import { messages } from '../messages';
import type { HandleSetViewState } from '../types';
import { Button } from './Button';

/**
 * Component showing "accept" or "reject" for Advertising purpose
 */
export function AcceptOrRejectAdvertising({
  handleSetViewState,
}: {
  /** Function to change viewState */
  handleSetViewState: HandleSetViewState;
}): JSX.Element {
  const { airgap } = useAirgap();
  const { formatMessage } = useIntl();

  return (
    <div className="column-content">
      <div>
        <div>
          <p className="text-title text-title-left">
            {formatMessage(messages.consentTitleAcceptOrRejectAdvertising)}
          </p>
        </div>
        <div>
          <p className="paragraph">
            {formatMessage(messages.acceptOrRejectAdvertisingDescription)}
          </p>
        </div>
      </div>
      <div className="accept-or-reject-all-button-row">
        <Button
          primaryText={formatMessage(messages.acceptAdvertising)}
          handleClick={(event) => {
            event.preventDefault();
            airgap.setConsent(event, { Advertising: true }, CONSENT_OPTIONS);
            handleSetViewState('close');
          }}
        />
        <Button
          primaryText={formatMessage(messages.rejectAdvertising)}
          handleClick={(event) => {
            event.preventDefault();
            airgap.setConsent(event, { Advertising: false }, CONSENT_OPTIONS);
            handleSetViewState('close');
          }}
        />
      </div>
    </div>
  );
}
