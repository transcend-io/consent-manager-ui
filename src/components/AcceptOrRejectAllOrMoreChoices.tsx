import { h, JSX } from 'preact';
import { useIntl } from 'react-intl';
import { useAirgap } from '../hooks';
import { messages } from '../messages';
import type { HandleSetViewState } from '../types';
import { Button } from './Button';

/**
 * Component showing "accept all" or "reject all" or "more choices"
 */
export function AcceptOrRejectAllOrMoreChoices({
  handleSetViewState,
}: {
  /** Function to change viewState */
  handleSetViewState: HandleSetViewState;
}): JSX.Element {
  const { airgap } = useAirgap();
  const { formatMessage } = useIntl();

  // Opt in to all purposes
  const handleAcceptAll:
    | JSX.MouseEventHandler<HTMLButtonElement>
    | undefined = (
    event: JSX.TargetedEvent<HTMLButtonElement, MouseEvent>,
  ): void => {
    event.preventDefault();
    airgap.optIn(event);
    handleSetViewState('close');
  };

  // Opt out of all non-essential purposes
  const handleRejectAll:
    | JSX.MouseEventHandler<HTMLButtonElement>
    | undefined = (
    event: JSX.TargetedEvent<HTMLButtonElement, MouseEvent>,
  ): void => {
    event.preventDefault();
    airgap.optOut(event);
    handleSetViewState('close');
  };

  // redirect to more choices
  const handleMoreChoices:
    | JSX.MouseEventHandler<HTMLButtonElement>
    | undefined = (
    event: JSX.TargetedEvent<HTMLButtonElement, MouseEvent>,
  ): void => {
    event.preventDefault();
    handleSetViewState('CompleteOptions');
  };

  return (
    <div className="column-content">
      <div>
        <div>
          <p className="text-title text-title-left">
            {formatMessage(messages.consentTitleAcceptAll)}
          </p>
        </div>
        <div>
          <p className="paragraph">
            {formatMessage(messages.acceptAllDescription)}
          </p>
        </div>
      </div>
      <div className="accept-or-reject-all-button-row">
        <Button
          primaryText={formatMessage(messages.acceptAllButtonPrimary)}
          handleClick={handleAcceptAll}
        />
        <Button
          primaryText={formatMessage(messages.rejectAllButtonPrimary)}
          handleClick={handleRejectAll}
        />
        <Button
          primaryText={formatMessage(messages.moreChoicesButtonPrimary)}
          handleClick={handleMoreChoices}
        />
      </div>
    </div>
  );
}
