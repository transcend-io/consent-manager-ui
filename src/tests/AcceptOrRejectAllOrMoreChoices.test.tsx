import '@testing-library/jest-dom';

import { h } from 'preact';
import { render, testWindow } from './utils/render';
import { test, expect, describe } from '@jest/globals';

import { AcceptOrRejectAllOrMoreChoices } from '../components/AcceptOrRejectAllOrMoreChoices';
import { fireEvent } from '@testing-library/preact';
import {
  MOCK_PURPOSES_OPTED_IN,
  MOCK_PURPOSES_OPTED_OUT,
  MOCK_TEMPLATE_VARIABLES,
} from './utils/constants';
import { init as initMockAirgap } from './utils/ag-mock';
import { messages } from '../messages';

describe('AcceptOrRejectAllOrMoreChoices', () => {
  test('matches snapshot', () => {
    const { snapshot } = render(
      <AcceptOrRejectAllOrMoreChoices
        handleSetViewState={() => null}
        globalUiVariables={MOCK_TEMPLATE_VARIABLES}
        moreChoicesViewState="CompleteOptions"
      />,
    );
    expect(snapshot).toMatchSnapshot();
  });

  test('button ordering is consistent', () => {
    initMockAirgap(MOCK_PURPOSES_OPTED_OUT);
    const { container } = render(
      <AcceptOrRejectAllOrMoreChoices
        handleSetViewState={() => null}
        globalUiVariables={MOCK_TEMPLATE_VARIABLES}
        moreChoicesViewState="CompleteOptions"
      />,
    );

    const domLabels = [...container.querySelectorAll('button')]?.map(
      (span) => span.children[0].childNodes.item(0).textContent,
    );

    expect(domLabels).toStrictEqual([
      messages.acceptAllButtonPrimary.defaultMessage,
      messages.rejectAllButtonPrimary.defaultMessage,
      messages.moreChoicesButtonPrimary.defaultMessage,
    ]);
  });

  test('submission affects stored consent (opted out)', () => {
    initMockAirgap(MOCK_PURPOSES_OPTED_OUT);
    const prevConsent = { ...testWindow.airgap.getConsent() };
    const { container } = render(
      <AcceptOrRejectAllOrMoreChoices
        handleSetViewState={() => null}
        globalUiVariables={MOCK_TEMPLATE_VARIABLES}
        moreChoicesViewState="CompleteOptions"
      />,
    );

    // Ensure submission reflects purpose selection in airgap
    const acceptButton = container.querySelector('button');
    if (acceptButton) fireEvent.click(acceptButton);

    const consent = { ...testWindow.airgap.getConsent() };

    Object.entries(MOCK_PURPOSES_OPTED_OUT).forEach(([, purpose]) => {
      expect(!!prevConsent.purposes?.[purpose.name]).toEqual(
        !consent.purposes?.[purpose.name],
      );
    });
    expect(prevConsent.confirmed).toEqual(false);
    expect(consent.confirmed).toEqual(true);
  });

  test('submission affects stored consent (opted in)', () => {
    initMockAirgap(MOCK_PURPOSES_OPTED_IN);
    const prevConsent = { ...testWindow.airgap.getConsent() };
    const { container } = render(
      <AcceptOrRejectAllOrMoreChoices
        handleSetViewState={() => null}
        globalUiVariables={MOCK_TEMPLATE_VARIABLES}
        moreChoicesViewState="CompleteOptions"
      />,
    );

    // Ensure submission reflects purpose selection in airgap
    const rejectButton = container.querySelector('button:nth-of-type(2)');
    if (rejectButton) fireEvent.click(rejectButton);

    const consent = { ...testWindow.airgap.getConsent() };

    Object.entries(MOCK_PURPOSES_OPTED_IN).forEach(([, purpose]) => {
      expect(!!prevConsent.purposes?.[purpose.name]).toEqual(
        !consent.purposes?.[purpose.name],
      );
    });
    expect(prevConsent.confirmed).toEqual(false);
    expect(consent.confirmed).toEqual(true);
  });

  test('that CompleteOptions is opened', () => {
    let calledArg;
    const { container } = render(
      <AcceptOrRejectAllOrMoreChoices
        handleSetViewState={(state) => {
          calledArg = state;
        }}
        globalUiVariables={MOCK_TEMPLATE_VARIABLES}
        moreChoicesViewState="CompleteOptions"
      />,
    );

    const moreChoicesButton = container.querySelector('button:nth-of-type(3)');
    if (moreChoicesButton) fireEvent.click(moreChoicesButton);

    expect(calledArg).toEqual('CompleteOptions');
  });

  test('that CompleteOptionsToggles is opened', () => {
    let calledArg;
    const { container } = render(
      <AcceptOrRejectAllOrMoreChoices
        handleSetViewState={(state) => {
          calledArg = state;
        }}
        globalUiVariables={MOCK_TEMPLATE_VARIABLES}
        moreChoicesViewState="CompleteOptionsToggles"
      />,
    );

    const moreChoicesButton = container.querySelector('button:nth-of-type(3)');
    if (moreChoicesButton) fireEvent.click(moreChoicesButton);

    expect(calledArg).toEqual('CompleteOptionsToggles');
  });
});
