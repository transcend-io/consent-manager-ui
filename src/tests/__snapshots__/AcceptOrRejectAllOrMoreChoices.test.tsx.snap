// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AcceptOrRejectAllOrMoreChoices matches snapshot 1`] = `
"<div>
  <div class="column-content">
    <div>
      <div>
        <p
          id="consent-dialog-title"
          role="heading"
          class="text-title text-title-left"
        >
          We use cookies
        </p>
      </div>
      <div>
        <p class="paragraph">
          By clicking “Accept all”, you agree to the storing of cookies on your
          device for functional, analytics, and advertising purposes.
        </p>
      </div>
    </div>
    <div class="accept-or-reject-all-button-row">
      <button class="button">
        <span class="button-base-text button-primary-text"
          >Accept all</span
        ></button
      ><button class="button" data-initialfocus="true">
        <span class="button-base-text button-primary-text"
          >Reject all</span
        ></button
      ><button class="button">
        <span class="button-base-text button-primary-text">More choices</span>
      </button>
    </div>
  </div>
</div>
"
`;
